[SPARK-16323] [SQL] Add IntegerDivide to avoid unnecessary cast

techaddict opened the pull request:
## What changes were proposed in this pull request?

Add IntegerDivide to avoid unnecessary cast

Before:

```
scala> spark.sql("select 6 div 3").explain(true)
...
== Analyzed Logical Plan ==
CAST((6 / 3) AS BIGINT): bigint
Project [cast((cast(6 as double) / cast(3 as double)) as bigint) AS
CAST((6 / 3) AS BIGINT)#5L]
+- OneRowRelation$
...
```

After:

```
scala> spark.sql("select 6 div 3").explain(true)
...
== Analyzed Logical Plan ==
(6 / 3): int
Project [(6 / 3) AS (6 / 3)#11]
+- OneRowRelation$
...
```
## How was this patch tested?

Existing Tests and added new ones

techaddict committed:
[SPARK-16323] [SQL] Add IntegerDivide to avoid unnecessary cast

Before:
```
scala> spark.sql("select 6 div 3").explain(true)
...
== Analyzed Logical Plan ==
CAST((6 / 3) AS BIGINT): bigint
Project [cast((cast(6 as double) / cast(3 as double)) as bigint) AS
CAST((6 / 3) AS BIGINT)#5L]
+- OneRowRelation$
...
```

After:
```
scala> spark.sql("select 6 div 3").explain(true)
...
== Analyzed Logical Plan ==
(6 / 3): int
Project [(6 / 3) AS (6 / 3)#11]
+- OneRowRelation$
...
```

techaddict committed:
Merge branch 'master' into SPARK-16323

techaddict commented:
cc: @cloud-fan

SparkQA commented:
**[Test build #61682 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61682/consoleFull)** for PR 14036 at commit [`e4e42c3`](https://github.com/apache/spark/commit/e4e42c35b0236dff6aedf6468a7d94f80bc6023b).

SparkQA commented:
**[Test build #61682 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61682/consoleFull)** for PR 14036 at commit [`e4e42c3`](https://github.com/apache/spark/commit/e4e42c35b0236dff6aedf6468a7d94f80bc6023b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class MapKeys(child: Expression)`
  * `case class MapValues(child: Expression)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61682/
Test FAILed.

techaddict committed:
fix docs

# Conflicts:
#
sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Funct
ionRegistry.scala
#	sql/core/src/main/scala/org/apache/spark/sql/functions.scala
#
sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.sca
la
#
sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.sca
la

SparkQA commented:
**[Test build #61686 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61686/consoleFull)** for PR 14036 at commit [`5db17fd`](https://github.com/apache/spark/commit/5db17fd2e05578bd152d9f2654d49aa4e07abe5b).

SparkQA commented:
**[Test build #61686 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61686/consoleFull)** for PR 14036 at commit [`5db17fd`](https://github.com/apache/spark/commit/5db17fd2e05578bd152d9f2654d49aa4e07abe5b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61686/
Test FAILed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:
```diff
@@ -234,6 +234,7 @@ object FunctionRegistry {
     expression[Subtract]("-"),
     expression[Multiply]("*"),
     expression[Divide]("/"),
+    expression[IntegerDivide]("div"),
```
does hive support this syntax? i.e. `div(4, 2)`

techaddict replied:
I don't think so.

lianhuiwang replied:
@cloud-fan yes, hive support div and / .

techaddict replied:
@lianhuiwang doing ```div(4,2)``` gives
```
hive> div(4, 2);
NoViableAltException(14@[])
	at org.apache.hadoop.hive.ql.parse.HiveParser.statement(HiveParser.java:1099)
	at org.apache.hadoop.hive.ql.parse.ParseDriver.parse(ParseDriver.java:204)
	at org.apache.hadoop.hive.ql.parse.ParseDriver.parse(ParseDriver.java:166)
	at org.apache.hadoop.hive.ql.Driver.compile(Driver.java:440)
	at org.apache.hadoop.hive.ql.Driver.compile(Driver.java:319)
	at org.apache.hadoop.hive.ql.Driver.compileInternal(Driver.java:1249)
	at org.apache.hadoop.hive.ql.Driver.runInternal(Driver.java:1295)
	at org.apache.hadoop.hive.ql.Driver.run(Driver.java:1178)
	at org.apache.hadoop.hive.ql.Driver.run(Driver.java:1166)
	at org.apache.hadoop.hive.cli.CliDriver.processLocalCmd(CliDriver.java:236)
	at org.apache.hadoop.hive.cli.CliDriver.processCmd(CliDriver.java:187)
	at org.apache.hadoop.hive.cli.CliDriver.processLine(CliDriver.java:403)
	at org.apache.hadoop.hive.cli.CliDriver.executeDriver(CliDriver.java:782)
	at org.apache.hadoop.hive.cli.CliDriver.run(CliDriver.java:721)
	at org.apache.hadoop.hive.cli.CliDriver.main(CliDriver.java:648)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.hadoop.util.RunJar.run(RunJar.java:221)
	at org.apache.hadoop.util.RunJar.main(RunJar.java:136)
FAILED: ParseException line 1:0 cannot recognize input near 'div' '(' '4'
```

lianhuiwang replied:
'select 4 div 2' is the right code.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -285,6 +284,75 @@ case class Divide(left: Expression, right: Expression)
 @ExpressionDescription(
+  usage = "a _FUNC_ b - Divides a by b.",
+  extended = "> SELECT 3 _FUNC_ 2;\n 1")
+case class IntegerDivide(left: Expression, right: Expression)
```
There are a lot of duplicated code between this class and `Divide`, is there any possibility we can abstract them?

techaddict replied:
Let me try doing that 👍

techaddict replied:
Done

techaddict committed:
replace symbol / with div for integer division

SparkQA commented:
**[Test build #61689 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61689/consoleFull)** for PR 14036 at commit [`faa2fd3`](https://github.com/apache/spark/commit/faa2fd3b3d8a91f46bac6260e3ddd9745cbce758).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61689/
Test FAILed.

techaddict committed:
add decimalMethod

SparkQA commented:
**[Test build #61692 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61692/consoleFull)** for PR 14036 at commit [`e30747a`](https://github.com/apache/spark/commit/e30747a9126e88e04fa4adc36c4b42d70245cb85).

SparkQA commented:
**[Test build #61692 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61692/consoleFull)** for PR 14036 at commit [`e30747a`](https://github.com/apache/spark/commit/e30747a9126e88e04fa4adc36c4b42d70245cb85).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61692/
Test FAILed.

techaddict committed:
use / in java for integer division

techaddict committed:
/

SparkQA commented:
**[Test build #61695 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61695/consoleFull)** for PR 14036 at commit [`054d54e`](https://github.com/apache/spark/commit/054d54e1dae7c372fb1c7ef5ee52f7d9ff6a619d).

SparkQA commented:
**[Test build #61695 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61695/consoleFull)** for PR 14036 at commit [`054d54e`](https://github.com/apache/spark/commit/054d54e1dae7c372fb1c7ef5ee52f7d9ff6a619d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61695/
Test PASSed.

techaddict committed:
Merge branch 'master' into SPARK-16323

techaddict committed:
Abstract out into DivisionArithmetic

SparkQA commented:
**[Test build #61706 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61706/consoleFull)** for PR 14036 at commit [`ff97457`](https://github.com/apache/spark/commit/ff9745776fcf97ff063dec0811762a3e0c4b1840).

SparkQA commented:
**[Test build #61706 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61706/consoleFull)** for PR 14036 at commit [`ff97457`](https://github.com/apache/spark/commit/ff9745776fcf97ff063dec0811762a3e0c4b1840).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class DivisionArithmetic extends BinaryArithmetic with NullIntolerant `
  * `case class Divide(left: Expression, right: Expression)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61706/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -237,54 +229,100 @@ case class Divide(left: Expression, right: Expression)
   }
 
+  // Used by doGenCode
+  def divide(eval1: ExprCode, eval2: ExprCode, javaType: String): String
```
mark as `protected`?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -237,54 +229,100 @@ case class Divide(left: Expression, right: Expression)
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     val eval1 = left.genCode(ctx)
     val eval2 = right.genCode(ctx)
-    val isZero = if (dataType.isInstanceOf[DecimalType]) {
```
why not just keep it? Then the 2 implementations can share the same codegen.

techaddict committed:
make a single isZero

techaddict committed:
make divide protected

techaddict committed:
small fix

techaddict commented:
@cloud-fan addressed all your comments 🎉

SparkQA commented:
**[Test build #61766 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61766/consoleFull)** for PR 14036 at commit [`e89ffc0`](https://github.com/apache/spark/commit/e89ffc06691d0a1cda037aa63e20cdae3df77793).

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/Column.scala:
```diff
@@ -732,6 +732,21 @@ class Column(protected[sql] val expr: Expression) extends Logging {
+   * @group expr_ops
+   * @since 2.1.0
+   */
+  def div (other: Any): Column = withExpr { IntegerDivide(expr, lit(other).expr) }
```
why can't we just use the normal / ?

rxin commented:
@techaddict why do we need to introduce a new user facing function for this?

SparkQA commented:
**[Test build #61766 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61766/consoleFull)** for PR 14036 at commit [`e89ffc0`](https://github.com/apache/spark/commit/e89ffc06691d0a1cda037aa63e20cdae3df77793).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61766/
Test PASSed.

cloud-fan commented:
@rxin , the `/` in hive/mysql means fraction division, e.g. `3 / 2` will get `1.5`. And the `div` is a special operator for integral division, `3 div 2` will get `1`.

In spark sql we implement `div` by casting the result of `/` to int, which is inefficient.

rxin commented:
OK I think it's fine to add it in SQL for compatibility but I wouldn't add functions in the DataFrame API for that.

techaddict committed:
Merge branch 'master' into SPARK-16323

techaddict committed:
remove div

SparkQA commented:
**[Test build #61857 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61857/consoleFull)** for PR 14036 at commit [`2c02370`](https://github.com/apache/spark/commit/2c023704a4573b20946158410897213a5617b24f).

SparkQA commented:
**[Test build #61857 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61857/consoleFull)** for PR 14036 at commit [`2c02370`](https://github.com/apache/spark/commit/2c023704a4573b20946158410897213a5617b24f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61857/
Test PASSed.

techaddict commented:
@rxin @cloud-fan done.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -249,11 +244,7 @@ case class Divide(left: Expression, right: Expression)
       s"${eval2.value} == 0"
     }
     val javaType = ctx.javaType(dataType)
-    val divide = if (dataType.isInstanceOf[DecimalType]) {
```
Does it already cover both fraction and integral division?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -277,14 +268,52 @@ case class Divide(left: Expression, right: Expression)
 }
 
 @ExpressionDescription(
+  usage = "a _FUNC_ b - Divides a by b.",
```
we should mention this is a fraction division, i.e. the parameter must be fraction type and the result is also fraction.

techaddict committed:
add better message

techaddict committed:
remove fn

SparkQA commented:
**[Test build #62093 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62093/consoleFull)** for PR 14036 at commit [`a461e35`](https://github.com/apache/spark/commit/a461e35af7be59fa555b4bbe176ac3e9ef4d08ef).

SparkQA commented:
**[Test build #62093 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62093/consoleFull)** for PR 14036 at commit [`a461e35`](https://github.com/apache/spark/commit/a461e35af7be59fa555b4bbe176ac3e9ef4d08ef).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62093/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -237,6 +229,9 @@ case class Divide(left: Expression, right: Expression)
   }
 
+  // Used by doGenCode
+  protected def divide(eval1: ExprCode, eval2: ExprCode, javaType: String): String
```
I don't think we need this abstraction. [this one](https://github.com/apache/spark/pull/14036/files#diff-1516b10738479bbe190fb4e239258473L252) already covers both fraction and integral

techaddict replied:
I did it on purpose. we can't call `$div` on `byte's` and plus if I try to call `value = value1 / value2;` for decimals, I get ```Binary numeric promotion not possible on types "org.apache.spark.sql.types.Decimal" and "org.apache.spark.sql.types.Decimal"```.

cloud-fan replied:
you can see from [this test](https://github.com/apache/spark/pull/14036/files#diff-f3fc3b985a8877ee493f1a12ddfb98d6L143) that the previous `Divide` expression can support all integral types and fraction types.

BTW looks there is no special handling for integral type in your PR, for fraction type:
```
if (dataType.isInstanceOf[DecimalType]) {
  s"${eval1.value}.$decimalMethod(${eval2.value})"
} else {
  s"($javaType)(${eval1.value} $symbol ${eval2.value})"
}
```
while the `symbol` is `/`, so we generate `($javaType)(${eval1.value} / ${eval2.value})` for double and float.

for integral type:
```
($javaType)(${eval1.value} $decimalMethod (${eval2.value}))
```
while the `decimalMethod` is `/`, so it's still generates `($javaType)(${eval1.value} / ${eval2.value})`

techaddict replied:
@cloud-fan yes but getting ```A method named "$div" is not declared in any enclosing class nor any supertype, nor through a static import``` in the updated pr for `Code generation of (2.0 / 1.0)`

techaddict committed:
use default divide

techaddict committed:
Merge branch 'master' into SPARK-16323

SparkQA commented:
**[Test build #62177 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62177/consoleFull)** for PR 14036 at commit [`dabaadb`](https://github.com/apache/spark/commit/dabaadb29e17a55b15e8d7399fac6a32af158ef9).

SparkQA commented:
**[Test build #62179 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62179/consoleFull)** for PR 14036 at commit [`d057546`](https://github.com/apache/spark/commit/d057546cd1132b4ece05df2a4a7d721ac65ca600).

SparkQA commented:
**[Test build #62176 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62176/consoleFull)** for PR 14036 at commit [`aae3194`](https://github.com/apache/spark/commit/aae31941aaa7adf456104ecca39bdcf16ce24791).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62176/
Test FAILed.

SparkQA commented:
**[Test build #62177 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62177/consoleFull)** for PR 14036 at commit [`dabaadb`](https://github.com/apache/spark/commit/dabaadb29e17a55b15e8d7399fac6a32af158ef9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62177/
Test FAILed.

SparkQA commented:
**[Test build #62179 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62179/consoleFull)** for PR 14036 at commit [`d057546`](https://github.com/apache/spark/commit/d057546cd1132b4ece05df2a4a7d721ac65ca600).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62179/
Test FAILed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -249,11 +241,12 @@ case class Divide(left: Expression, right: Expression)
       s"${eval1.value}.$decimalMethod(${eval2.value})"
     } else {
-      s"($javaType)(${eval1.value} $symbol ${eval2.value})"
+      s"($javaType)(${eval1.value} $decimalMethod ${eval2.value})"
```
ok I see the problem. Since `symbol` is also used in the string format of arithmetic expression, `IntegerDivide` has to override it.

I think we can:

1. inline the operator in generated code, e.g. `($javaType)(${eval1.value} / ${eval2.value})` instead of `($javaType)(${eval1.value} $symbol ${eval2.value})`
2. don't override `decimalMethod` as it's useless in the 2 division expressions.

techaddict replied:
but what about the decimalMethod used in line 245 ? if we inline the operator `/` there too it gives 
```Binary numeric promotion not possible on types "org.apache.spark.sql.types.Decimal" and "org.apache.spark.sql.types.Decimal"```

cloud-fan replied:
just follow the previous behaviour, inline "$div"

techaddict committed:
Merge branch 'master' into SPARK-16323

techaddict committed:
fix

techaddict committed:
private lazy

SparkQA commented:
**[Test build #62212 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62212/consoleFull)** for PR 14036 at commit [`ab6858c`](https://github.com/apache/spark/commit/ab6858cac3f8f53a3437038b7cd767e73d170eaa).

techaddict committed:
remove decimalMethod and do $$div

techaddict commented:
@cloud-fan Updated the PR, all tests should pass now.

SparkQA commented:
**[Test build #62213 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62213/consoleFull)** for PR 14036 at commit [`8d9a04d`](https://github.com/apache/spark/commit/8d9a04d61a155f5bc131cc7a06a1f9378ceb1cbe).

SparkQA commented:
**[Test build #62212 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62212/consoleFull)** for PR 14036 at commit [`ab6858c`](https://github.com/apache/spark/commit/ab6858cac3f8f53a3437038b7cd767e73d170eaa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62212/
Test PASSed.

SparkQA commented:
**[Test build #62213 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62213/consoleFull)** for PR 14036 at commit [`8d9a04d`](https://github.com/apache/spark/commit/8d9a04d61a155f5bc131cc7a06a1f9378ceb1cbe).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62213/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -285,6 +278,28 @@ case class Divide(left: Expression, right: Expression)
+@ExpressionDescription(
+  usage = "a _FUNC_ b - Divides a by b.",
+  extended = "> SELECT 3 _FUNC_ 2;\n 1")
+case class IntegerDivide(left: Expression, right: Expression)
```
`IntegralDivide`?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -207,20 +207,12 @@ case class Multiply(left: Expression, right: Expression)
-
-  override def symbol: String = "/"
-  override def decimalMethod: String = "$div"
+abstract class DivisionArithmetic extends BinaryArithmetic with NullIntolerant {
```
how about `DivideBase`?

cloud-fan commented:
LGTM except 2 naming comments, thanks for working on it!

techaddict committed:
Merge branch 'master' into SPARK-16323

techaddict committed:
Rename

techaddict committed:
remove whitespace

techaddict committed:
style fix

techaddict commented:
@cloud-fan Done 👍

SparkQA commented:
**[Test build #62224 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62224/consoleFull)** for PR 14036 at commit [`16eff20`](https://github.com/apache/spark/commit/16eff2071a1ce2f532000e61f6990eb9d77c173f).

SparkQA commented:
**[Test build #62224 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62224/consoleFull)** for PR 14036 at commit [`16eff20`](https://github.com/apache/spark/commit/16eff2071a1ce2f532000e61f6990eb9d77c173f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Divide(left: Expression, right: Expression) extends DivideBase `
  * `case class IntegralDivide(left: Expression, right: Expression) extends DivideBase `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62224/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -285,6 +278,26 @@ case class Divide(left: Expression, right: Expression)
 }
 
 @ExpressionDescription(
+  usage = "a _FUNC_ b - Fraction Division a by b.",
```
how about `Divides a by b of fraction type`?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -285,6 +278,26 @@ case class Divide(left: Expression, right: Expression)
+}
+
+@ExpressionDescription(
+  usage = "a _FUNC_ b - Divides a by b.",
```
how about Divides a by b of integral type?

cloud-fan commented:
last 2 comments. @liancheng @clockfly  can you also take a look?

lianhuiwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -957,7 +957,7 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
         Remainder(left, right)
       case SqlBaseParser.DIV =>
-        Cast(Divide(left, right), LongType)
+        IntegralDivide(left, right)
```
I think we needs to add SqlBaseParser.DIVIDE for '/'.

lianhuiwang replied:
That's ok because I find SparkSQL has SqlBaseParser.SLASH  for '/' .

lianhuiwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -207,20 +207,12 @@ case class Multiply(left: Expression, right: Expression)
 
   private lazy val div: (Any, Any) => Any = dataType match {
     case ft: FractionalType => ft.fractional.asInstanceOf[Fractional[Any]].div
+    case i: IntegralType => i.integral.asInstanceOf[Integral[Any]].quot
```
@cloud-fan how about make this line in IntegralDivide that can be more readable?

cloud-fan replied:
The `DivideBase` implement codegen for all types, so I think it's fine for it to implement `eval` for all types.

lianhuiwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:
```diff
@@ -207,20 +207,12 @@ case class Multiply(left: Expression, right: Expression)
-
-  override def inputType: AbstractDataType = TypeCollection(DoubleType, DecimalType)
-
-  override def symbol: String = "/"
```
we need to keep symbol because it is very useful for BinaryOperator' sqlOperator() and sql()

yhuai commented:
@techaddict Can you test the performance with and without your change?

techaddict commented:
@yhuai sure, doing performance testing using sql query or expression ?

yhuai commented:
Having a query just to test this expression is good.

HyukjinKwon commented:
@techaddict How is the perf test going?

techaddict commented:
@HyukjinKwon didn't have bandwidth will try to finish this weekend
